lms/djangoapps/courseware/courses.py

```diff
@@ -449,10 +449,44 @@ def get_course_syllabus_section(course, section_key):
     raise KeyError("Invalid about key " + str(section_key))
 
 
+class AccessLimitedCourseList(object):
+    def __init__(self, course_queryset, user, permission_name):
+        self.course_queryset = course_queryset
+        self._access_limited_courses = (c for c in self.course_queryset if has_access(user, permission_name, c))
+        self._count = None
+        self._course_cache = []
+
+    def __len__(self):
+        if self._count is None:
+            self._count = self.course_queryset.count()
+        return self._count
+
+    def __iter__(self):
+        for course in self._course_cache:
+            yield course
+        for course in self._access_limited_courses:
+            self._course_cache.append(course)
+            yield course
+
+    def __getitem__(self, index):
```

Comment:
all the methods in this class but especially this method could use some comments.  
I think the most helpful thing would be a high level comment explaining what the class is doing and then a comment that includes a brief example of what each method would do with a particular course  
with a comment I may be able to better understand this method and review it, but just based on the name i wonder if this kind of thing could somehow be more simply defined with some built in django functions